### PR TITLE
Initial operator changes implementation

### LINKED ISSUE #16730: Should operators be defined as methods on types instead of stand-alone functions?
This has come to mind as a result of the questions posed by #16725 and #16727, but we have toyed with it in passing before.

Checking a subset of languages that allow operator overloading, it looks like:

- C++ allows defining operators as both methods (`R K::operator +(S b);` within the type definition of K) and stand-alone functions (`R operator +(K a, S b);`.
- Python allows defining operators as methods (`__add__(self, other)`)
- Rust has a list of traits that expect method definitions within their context.  So you connect the trait to the type and then define the necessary method (see [this link](https://doc.rust-lang.org/rust-by-example/trait/ops.html) for a better explanation)
- Near as I can tell, Julia allows defining operators as functions (`Base.:+(x::SomeType, y::SomeType) = ...`)
- It seems like D allows defining operators as methods, though the docs says "functions" - it may be I don't know enough about D and functions and methods are not distinguishable there [(see this link)](https://dlang.org/spec/operatoroverloading.html)

Pros:
- Ties the operator to the type(s) it operates on more closely
- Would make the implementation to access operators defined in the same location as a type when the type's scope is not otherwise visible trivial (as we already resolve methods by jumping to that scope)

Cons:
- For operators that take multiple arguments, choosing which one to define it on is kinda arbitrary and can lead to confusion when trying to access operators defined outside the scope where the type is defined
- Will require changing a lot of code
- Special handling to treat them similarly to methods in the compiler today would be replaced by special handling to not treat them similarly in other cases (for instance, it seems like we wouldn't want forwarding to apply to operators, see #11717)

### LINKED ISSUE #16929: Proposal: "operator" keyword for operator functions and methods
This proposal is a slightly more polished version of the discussion that occurred via email and starting from [this comment](https://github.com/chapel-lang/chapel/issues/16730#issuecomment-748289279), as originally proposed by Engin and discussed with Brad and Michael.  It is intended to bring more eyes to the idea by giving it its own topic, and hopefully reach a final decision.

Operators have traditionally been implemented as stand-alone functions, distinguishable from regular functions largely by their name ("+", etc.).  We're moving towards also allowing operators to be explicitly associated with the types on which they are defined (again, see discussion on #16730), and expect to continue to do so.  This issue proposes that we also move to make operators more explicitly distinguishable from other functions and methods, by using a new keyword `operator`.  E.g., rather than writing:

```chapel
record R {
  // primary operator
  proc type +(lhs: R, rhs: R) { ... }
}

// secondary operator
proc type R.-(lhs: R, rhs: R) { ... }

// standalone operator
proc *(lhs: R, rhs: S) { ... }
```

We would instead write:

```chapel
record R {
  // primary operator
  operator +(lhs: R, rhs: R) { ... }
}

// secondary operator
operator R.-(lhs: R, rhs: R) { ... }

// standalone operator
operator *(lhs: R, rhs: S) { ... }
```

This would make a special cased style of function more explicitly recognizable.  It would be great as part of the implementation to move some of the compiler special-casing for operators to be triggered by the operator keyword (but that's an implementation detail).

Note that I would expect this to be a breaking change.  If we do not immediately drop support for operators without the keyword, I would expect us to do so within a few releases.

Names
-------

As part of this issue, I propose at least initially limiting the names used for operators to only those already known to the Chapel language.  For instance, the following definitions would be allowed because Chapel already is aware of the "+" and "-" operators:

```chapel
operator +(lhs: R, rhs: R) { ... }
operator R.-(lhs: R, rhs: R) { ... }
```

but the following definition would cause a compilation error, because Chapel is not aware of an operator "$%&":

```chapel
operator $%&(lhs: R, rhs: R) { ... }
```

We can choose to allow the operator keyword to define new operators in the future, but I'd rather we stick to supporting what is already supported for now.  If we are able to tie operator functionality more generally to the keyword, that could make such support more feasible in the future.

Operator Iterators
------------------

Though today it is possible to define an operator as an iterator, in our discussions we found this support surprising and unintentional.  I propose that we choose not to allow such cases until such time as they are explicitly requested.  Engin suggested in [this comment](https://github.com/chapel-lang/chapel/issues/16730#issuecomment-754113978) that if we decide to allow operator iterators, we make the syntax `operator iterator` and allow  `operator proc`, but assume that `operator` without either defaults to `operator proc`.  I think this is a good plan.

### PR DESCRIPTION
**Pull request by lydia-duncan**
Adds support for operator methods, as well as declaring operator functions
and methods with the new "operator" keyword.

Resolves #16730 
Resolves #16929 

<details>
Adds `operator` as a keyword to the parser, usable when declaring a function in
place of `proc` or `iter`.

Adds a new flag, FLAG_OPERATOR, which is explicitly applied to functions and
methods declared with the operator keyword and implicitly applied to functions
with recognized names for now.  The place where the flag is attached to functions
with particular names is a good place to add a deprecation warning (but that will
be handled in a later PR).

When an operator is declared, we also will set its this intent tag to INTENT_TYPE,
so that if it does end up being a method, it will be recognized as a type method
and handled appropriately (by removing the "this" arg and its corresponding
actual for all calls after resolution).

Removes a check in normalize that prevents operators from being methods.

Adjusts several stages of function resolution to handle operator methods.  Allows
calls to operators to either contain or not contain the method token and "this"
actual, and allows operator functions to either contain or not contain the
corresponding formals.  This involves skipping those actuals or formals when
checking if a function is a match and when determining which of two
functions is a more specific match.  It also involves adjusting the call appropriately
when a match is found if extra actuals are present or if certain actuals are missing.

During disambiguation, pulled common code out of testArgMapping and into a
helper function - this had the happy effect of also fixing an error where we were
passing the wrong formal and function to canDispatch (though it seems as though
there was not a major impact from doing so)
</details>

Adds a number of tests to lock in the expected behavior, including:
- primary, secondary and tertiary operator methods
- methods on classes and methods on records
- operators where the arguments are the generic version of the type
- operators where the arguments are a particular instantiation of the generic type
- operators when inheritance is involved (currently a future, will be handled later)
- case where two types define an identical operator as a method (and so are in conflict)
- case where the operator is defined a type that matches the second argument but not
the first
- case where an operator has been defined as a method that conflicts with a operator
defined as a standalone function
- case where an operator has been defined with the same name as a standalone operator
but the two are not in conflict
- case where a type has an operator defined as a method and another operator defined as
a standalone function
- operator keyword applied to a standalone operator function

Removes two tests that were locking in that operators could not be methods.  Also updates
a future requesting operator methods for the current vision for the feature and resolves that
future.

Passed a full paratest with futures

TODO:
- [x] Commit parser changes after review

### CODE REVIEW

**Comment by lydia-duncan** [comment]
Started looking at labeling old operators with the keyword and discovered a case I'd missed, stand by

**Review by mppf — COMMENTED** [review]
I have some initial feedback but I need to finish looking at functionResolution.cpp and wrapper.cpp changes.

File: compiler/parser/chapel.ypp [hunk 1]
```diff
@@ -1424,12 +1425,23 @@ fn_decl_stmt:
...
           USR_FATAL_CONT(fn, "'iter' is not legal with 'extern'");
         fn->addFlag(FLAG_ITERATOR_FN);
       }
+      if ($2 == ProcIterOp_OP) {
+        fn->addFlag(FLAG_OPERATOR);
+        if (fn->_this != NULL) {
+          if (fn->thisTag == INTENT_BLANK) {
+            fn->thisTag = INTENT_TYPE;
+          } else {
+            badOperatorThisTagError(fn->name);
```
[the remark this answers is not in the document]
Comment (lydia-duncan) [reply compiler/parser/chapel.ypp:1440]: I think so - this handles secondary and tertiary methods, while the addDeclarations code handles primary methods.  I'll look into combining them, thanks!

File: compiler/passes/normalize.cpp [hunk 2]
```diff
@@ -255,12 +255,6 @@ void normalize() {
         fn->name = astrDeinit;
       }
 
-    // make sure methods don't attempt to overload operators
-    } else if (isalpha(fn->name[0])         == 0   &&
```
[the remark this answers is not in the document]
Comment (lydia-duncan) [reply compiler/passes/normalize.cpp:259]: That case is already handled more generally for all functions (not just methods) during parsing.  So I think I'll leave this code as is?

[the remark this answers is not in the document]
Comment (mppf) [reply compiler/passes/normalize.cpp:259]: Sure.

File: compiler/resolution/functionResolution.cpp [hunk 3]
```diff
@@ -5643,6 +5659,53 @@ static void testArgMapping(FnSymbol*                    fn1,
...
+    EXPLAIN("Function 2 is an operator standalone function");
+
+    Type* f1Type = formal1->type->getValType();
+    bool formal1Promotes = false;
+    bool formal1Narrows = false;
+
+    testArgMapHelper(fn1, formal1, actual, f1Type, actualType, actualParam,
+                     &formal1Promotes, &formal1Narrows, DC, DS, 1);
+
+    DS.fn1WeakPreferred = true;
```
Comment (mppf) [inline compiler/resolution/functionResolution.cpp:5704]: You'll need to update the section of the spec about function disambiguation.

**Review by mppf — APPROVED** [review]

**Comment by bradcray** [comment]
Woohoo!

**Pull request merged**